langchain4j-ark/src/main/java/dev/langchain4j/model/ark/ArkChatModel.java

```diff
@@ -0,0 +1,123 @@
+public class ArkChatModel implements ChatLanguageModel {
+    private final String apiKey;
+    private final String model;
+    private final Double topP;
+    private final Double frequencyPenalty;
+    private final Double presencePenalty;
+    private final Double temperature;
+    private final List<String> stops;
+    private final Integer maxTokens;
+    private final String user;
+    private final ArkService service;
+
+    @Builder
+    protected ArkChatModel(String apiKey,
+                           String model,
+                           Double topP,
+                           Double frequencyPenalty,
+                           Double presencePenalty,
+                           Double temperature,
+                           List<String> stops,
+                           Integer maxTokens,
+                           String user) {
+        if (isNullOrBlank(apiKey)) {
+            throw new IllegalArgumentException("Ark api key must be defined. It can be generated here: https://www.volcengine.com/docs/82379/1263279");
+        }
+        if (isNullOrBlank(model)) {
+            throw new IllegalArgumentException("Ark model(endpoint_id) must be defined. ");
+        }
+        this.apiKey = apiKey;
+        this.model = model;
+        this.topP = topP;
+        this.frequencyPenalty = frequencyPenalty;
+        this.presencePenalty = presencePenalty;
+        this.temperature = temperature;
+        this.stops = stops;
+        this.maxTokens = maxTokens;
+        this.user = user;
+        this.service = new ArkService(apiKey);
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages) {
+        return generateModel(messages, null, null);
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
+        return generateModel(messages, toolSpecifications, null);
+    }
+
+    private Response<AiMessage> generateModel(
+            List<ChatMessage> messages,
+            List<ToolSpecification> toolSpecifications,
+            ToolSpecification toolThatMustBeExecuted
+    ) {
+        ensureNotEmpty(messages, "messages");
+        ChatCompletionRequest.Builder builder = ChatCompletionRequest.builder()
+                .model(model)
+                .topP(topP)
+                .frequencyPenalty(frequencyPenalty)
+                .presencePenalty(presencePenalty)
+                .temperature(temperature)
+                .maxTokens(maxTokens)
+                .user(user)
+                .messages(toArkMessages(messages));
+        if (stops != null) {
+            builder.stop(stops);
+        }
+        if (toolSpecifications != null && !toolSpecifications.isEmpty()) {
+            builder.tools(toToolFunctions(toolSpecifications));
+        }
+        ChatCompletionResult chatCompletionResult = service.createChatCompletion(builder.build());
+
+        // shutdown service
+        service.shutdownExecutor();
```

Comment:
Question:
The overhead of frequently creating and destroying thread pools: When ArkService is created, an OkHttpClient is created. The dispatcher of this OkHttpClient maintains an ExecutorService thread pool internally to execute HTTP requests.By default, this ExecutorService is a ThreadPoolExecutor.
OkHttpClient It is designed as a shared client that runs for long periods of time, and it is generally not recommended to create and destroy frequently.Between multiple requests, OkHttpClient The thread pool can be reused.If the thread pool is frequently destroyed, it will lead to performance degradation.
Delay of thread pool shutdown: The shutdownExecutor() method only sends a signal to shut down the thread pool. The threads in the thread pool will not terminate immediately, but will wait until all tasks currently being executed have completed before terminating.If there are long-running tasks, the thread pool may remain active for a longer period of time.
 
Suggestion:
Thread pool reuse: Consider reusing the thread pools in ArkService instances, such as singleton ArkService, delayed initialization and shutdown, or handing them over to Spring for management.Also, check whether the executorService is null before trying to close it.

Reply:
这个咱们交流下啊
我看了下ArkService，他有apiKey、aksk、arkapi及组合timeout等方式创建，不太方便区分
依赖spring也不太好
shutdownExecutor方法也没有销毁model时调用的契机
然后我发现public ArkService(ArkApi api, ExecutorService executorService)，也许可以共享ExecutorService，但不依赖spring我还想不到如何管理他的生命周期，只能交给使用者自己去管理

我并没太多这方面经验，帮我看看我描述的有没有问题，更好的实现方式是什么？

Reply:
1. Thread Pool Management
Issue: The thread pool in ArkService is managed by the Dispatcher of OkHttpClient, but the maximum number of threads, maximum requests per host, and other parameters are not configured. If the request volume increases and the thread pool is not properly shut down, it could lead to unbounded growth of the thread pool, increasing memory usage, and potentially causing an OutOfMemoryError.
Suggestion:
When creating OkHttpClient, explicitly set the Dispatcher parameters such as setMaxRequests and setMaxRequestsPerHost to control the size of the thread pool.
Consider configuring a reasonable idle thread retention time using ConnectionPool to control the time threads remain idle and the number of connections, thereby reducing resource consumption.
2. ArkService Instance Management
Issue: An instance of ArkService is created each time in ArkChatModel, which could lead to resource waste, especially in scenarios with frequent calls.
Suggestion: Modify the instantiation logic of ArkService to a singleton pattern (or another suitable instance management strategy) to ensure that only one ArkService instance is created throughout the lifecycle of ArkChatModel, thereby saving resources.
3. Thread Pool Shutdown
Issue: The ArkService class provides a shutdownExecutor method to shut down the thread pool, but in the original code, there is no explicit call to this method. This could lead to threads in the thread pool occupying resources even when the service is no longer in use.
Suggestion: Add a lifecycle management method, such as shutdown, in ArkChatModel to manually call shutdownExecutor when ArkService is no longer needed, releasing resources. Ensure this method is called in appropriate places, such as when the application exits.ArkChatModel supports external input of thread pool construction parameters and input of thread pools, but will this violate the principles of langchain4j?
4. Efficient Use of Resources
Issue: ArkService creates a new OkHttpClient with each call, although OkHttpClient is thread-safe and can be reused. This could lead to unnecessary resource overhead.
Suggestion: Consider making OkHttpClient a singleton or shared instance, especially when it’s used with the same configuration repeatedly, thereby reducing unnecessary overhead.
﻿
@noday @langchain4j

Reply:
I would not worry too much about it, we are going to replace okhttp with jdk11 http client pretty soon